src/main/kotlin/frc/robot/subsystems/vision/VisionIOPhotonVision.kt

```diff
@@ -0,0 +1,143 @@
+            } else inputs.latestTargetObservation = TargetObservation(Rotation2d(), Rotation2d())
+
+
+            // Update PhotonPoseEstimator based on gyro readings
+            localPoseEstimator.addHeadingData(result.timestampSeconds, botRotation())
+
+            // Add pose observation
+            if (result.multitagResult.isPresent) { // Multitag result
+                val multitagResult = result.multitagResult.get()
+
+                // Calculate robot pose
+                val fieldToCamera = multitagResult.estimatedPose.best
+                val fieldToRobot = fieldToCamera.plus(robotToCamera().inverse())
+                val robotPose = Pose3d(fieldToRobot.translation, fieldToRobot.rotation)
+
+                // Calculate average tag distance
+                val totalTagDistance = result.targets.sumOf {
+                    it.bestCameraToTarget.translation.norm
+                }
+
+                // Add tag IDs
+                tagIds.addAll(multitagResult.fiducialIDsUsed)
+
+                // Add observation
+                poseObservations.add(
+                    PoseObservation(
+                        result.timestampSeconds,
+                        robotPose,
+                        multitagResult.estimatedPose.ambiguity,
+                        multitagResult.fiducialIDsUsed.size,
+                        totalTagDistance / result.targets.size,
+                        PoseObservationType.PHOTONVISION
+                    )
+                )
+            } else if (result.targets.isNotEmpty()) { // Single tag result
+                val target = result.targets[0]
+
+                // Calculate robot pose
+                aprilTagLayout.getTagPose(target.fiducialId).ifPresent { tagPose ->
+                    val fieldToTarget = Transform3d(tagPose.translation, tagPose.rotation)
+                    val cameraToTarget = target.bestCameraToTarget
+                    val fieldToCamera = fieldToTarget.plus(cameraToTarget.inverse())
+                    val fieldToRobot = fieldToCamera.plus(robotToCamera().inverse())
+                    val robotPose = Pose3d(fieldToRobot.translation, fieldToRobot.rotation)
+
+                    // Add tag ID
+                    tagIds.add(target.fiducialId.toShort())
+
+                    // Add observation
+                    poseObservations.add(
+                        PoseObservation(
+                            result.timestampSeconds,
+                            robotPose,
+                            target.poseAmbiguity,
+                            1,
+                            cameraToTarget.translation.norm,
+                            PoseObservationType.PHOTONVISION
+                        )
+                    )
+                }
+            }
+        }
+
+        // Save pose observations and tag IDs to inputs object
+        inputs.poseObservations = poseObservations.toTypedArray()
+        inputs.tagIds = tagIds.map { it.toInt() }.toTypedArray()
```

Comment:
```suggestion
        inputs.tagIds = tagIds.map { it.toInt() }.toIntArray()
```
This is more efficient, perhaps this can be changed in other places.